<a href="https://colab.research.google.com/github/YoshiyukiKono/langchain_for_beginners/blob/main/1_RAG_table_text_hugging-face-embeddings.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# LangChain DBテーブルのテキストデータを使ったRAG

https://api.python.langchain.com/en/latest/vectorstores/langchain.vectorstores.cassandra.Cassandra.html#

## パッケージインストール

In [3]:
!pip install openai==1.3.5 tiktoken==0.5.1 cohere==4.36
!pip install langchain==0.1.0

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 220.8/220.8 kB 2.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 15.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 48.9/48.9 kB 6.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 75.9/75.9 kB 10.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.7/2.7 MB 28.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 76.9/76.9 kB 9.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.3/58.3 kB 6.9 MB/s eta 0:00:00
  Attempting uninstall: importlib_metadata
    Found existing installation: importlib-metadata 7.0.1
    Uninstalling importlib-metadata-7.0.1:
      Successfully uninstalled importlib-metadata-7.0.1


In [2]:
import langchain

print(langchain.__version__)

0.1.0


In [4]:
!pip install cassandra-driver==3.28.0 cassio==0.1.3

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.1/19.1 MB 2.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.1/40.1 kB 2.0 MB/s eta 0:00:00


## Astra

Connect Bundleファイルを実行環境に配置します。

Connect Bundleファイルを Astra から Colab 環境に直接ダウンロードすることもできます。 あなたのAstra DB環境からコピーした情報で下のセルを変更してください。「WGET」タブからコマンド全体をコピーすることができるため、ノートブックで実行する際に、先頭に!を補って、そのまま用いることができます。

ただし、このAstraコントロールプレーンに表示されるURL は静的ではないことに注意が必要です。そのため、後日、再度実行する場合には、URL を再度コピーし直す必要があります。

あるいは、AstraDBコントロールプレーンからダウンロードしたファイルを左サイドメニュー「Files」からアップロードしてください。

In [ ]:
!wget -O secure-connect-demo.zip "https://..." <-  左のパスを消して、自分のパスをコピーしてください。

--2023-09-01 05:03:54--  https://datastax-cluster-config-prod.s3.us-east-2.amazonaws.com/d5556151-ea9a-4309-8be3-b8ea2b1cd03d-1/secure-connect-demo.zip?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIA2AIQRQ76S2JCB77W%2F20230901%2Fus-east-2%2Fs3%2Faws4_request&X-Amz-Date=20230901T050340Z&X-Amz-Expires=300&X-Amz-SignedHeaders=host&X-Amz-Signature=7db1257a20828e206535d85e46ffdfe26c4fe1166781a157033bbaac549b23c3
Resolving datastax-cluster-config-prod.s3.us-east-2.amazonaws.com (datastax-cluster-config-prod.s3.us-east-2.amazonaws.com)... 3.5.131.128, 52.219.95.34, 52.219.179.34, ...
Connecting to datastax-cluster-config-prod.s3.us-east-2.amazonaws.com (datastax-cluster-config-prod.s3.us-east-2.amazonaws.com)|3.5.131.128|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 12247 (12K) [application/zip]
Saving to: ‘secure-connect-demo.zip’

secure-connect-demo 100%[===================>]  11.96K  --.-KB/s    in 0s      

2023-09-01 05:03:54 (119 MB/s) - ‘secure-conne

In [6]:
!wget -O secure-connect-demo.zip "https://datastax-cluster-config-prod.s3.us-east-2.amazonaws.com/2c170f07-40f1-4550-b361-6b2c8dabb94f-1/secure-connect-demo.zip?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIA2AIQRQ76S2JCB77W%2F20240111%2Fus-east-2%2Fs3%2Faws4_request&X-Amz-Date=20240111T004137Z&X-Amz-Expires=300&X-Amz-SignedHeaders=host&X-Amz-Signature=441043aa58cd3943b40d3cc0c32ef9d2c14f8092013302451f2c90c94459a7b1"

--2024-01-11 00:41:53--  https://datastax-cluster-config-prod.s3.us-east-2.amazonaws.com/2c170f07-40f1-4550-b361-6b2c8dabb94f-1/secure-connect-demo.zip?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIA2AIQRQ76S2JCB77W%2F20240111%2Fus-east-2%2Fs3%2Faws4_request&X-Amz-Date=20240111T004137Z&X-Amz-Expires=300&X-Amz-SignedHeaders=host&X-Amz-Signature=441043aa58cd3943b40d3cc0c32ef9d2c14f8092013302451f2c90c94459a7b1
Resolving datastax-cluster-config-prod.s3.us-east-2.amazonaws.com (datastax-cluster-config-prod.s3.us-east-2.amazonaws.com)... 16.12.65.146, 52.219.97.202, 52.219.143.66, ...
Connecting to datastax-cluster-config-prod.s3.us-east-2.amazonaws.com (datastax-cluster-config-prod.s3.us-east-2.amazonaws.com)|16.12.65.146|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 12254 (12K) [application/zip]
Saving to: ‘secure-connect-demo.zip’

secure-connect-demo 100%[===================>]  11.97K  --.-KB/s    in 0s      

2024-01-11 00:41:53 (165 MB/s) - ‘secure-co

In [7]:
SECURE_CONNECT_BUNDLE_PATH = 'secure-connect-demo.zip'

In [8]:
!ls -l $SECURE_CONNECT_BUNDLE_PATH

-rw-r--r-- 1 root root 12254 Nov 30 00:08 secure-connect-demo.zip


ダウンロードやアップロードに失敗し、サイズゼロのファイルが作成されることがあります。以下のように、ファイルサイズが表示されていることを確認ください

-rw-r--r-- 1 root root **12254** Nov 30 00:08 secure-connect-demo.zip

In [9]:
import getpass

ASTRA_CLIENT_ID = getpass.getpass()

··········


In [10]:

ASTRA_CLIENT_SECRET = getpass.getpass()

··········


In [11]:
from cassandra.cluster import Cluster
from cassandra.auth import PlainTextAuthProvider

cloud_config= {
  'secure_connect_bundle': SECURE_CONNECT_BUNDLE_PATH
}
auth_provider = PlainTextAuthProvider(ASTRA_CLIENT_ID, ASTRA_CLIENT_SECRET)
cluster = Cluster(cloud=cloud_config, auth_provider=auth_provider)
session = cluster.connect()

row = session.execute("select release_version from system.local").one()
if row:
  print(row[0])
else:
  print("An error occurred.")

ERROR:cassandra.connection:Closing connection <AsyncoreConnection(132113045258736) 2c170f07-40f1-4550-b361-6b2c8dabb94f-us-east1.db.astra.datastax.com:29042:74b9b6e7-a3c6-4136-bd6d-c091614fc40d> due to protocol error: Error from server: code=000a [Protocol error] message="Beta version of the protocol used (5/v5-beta), but USE_BETA flag is unset"


4.0.11-c44bc828c0c8


In [12]:
YOUR_KEYSPACE = 'langchain'
session.set_keyspace(YOUR_KEYSPACE)
session

### DATA

In [13]:
%%writefile books.csv
ID,title,author,publisher,price,year,description
001,ネットワークセキュリティ入門,山田太郎,技術書典,4500,2021,ネットワークセキュリティの基本原則と実践的な手法に焦点を当てた入門書。
002,ビジネスデータ分析の基礎,田中花子,ビジネスブックス,3800,2020,ビジネスにおけるデータ分析の基本的な手法とツールの使い方を解説。
003,クラウドアーキテクチャデザイン,鈴木次郎,技術出版株式会社,5500,2022,クラウドネイティブなアプリケーションのアーキテクチャデザインとベストプラクティスを詳細に解説。
004,金融取引のアルゴリズム,林美香,データ出版社,5200,2019,金融市場におけるアルゴリズムトレーディングの基本原則と実践手法に焦点を当てた実践的なガイド。
005,人事マネジメントの実践,佐藤健太,ビジネスパートナーズ,4800,2020,組織内の人材マネジメントに関する基本的な原則と実践的な手法を紹介。
006,自然言語処理入門,伊藤さやか,技術書典,4200,2021,自然言語処理の基本概念から、実践的なテキストデータ処理手法までを解説。
007,マーケティング戦略の立案,山口直樹,ビジネスブックス,3600,2018,市場分析、セグメンテーション、ポジショニングなど、マーケティング戦略の構築手法を学ぶ。
008,ディープラーニングアプリケーション,田村徹,技術出版株式会社,6500,2019,ディープラーニングを使用した実践的なアプリケーション開発手法に焦点を当てた実践書。
009,投資ポートフォリオ管理,高橋優一,金融情報出版,3700,2020,効果的な投資ポートフォリオの構築と管理に関する手法を解説。
010,サイバーセキュリティ実践ガイド,中村あや,技術書典,4800,2021,サイバーセキュリティの基本から、脆弱性診断、セキュリティ対策までを詳細に解説。
011,ビジネスリーダーシップの原則,大塚和夫,ビジネスパートナーズ,4200,2020,ビジネスにおけるリーダーシップの基本原則と実践的なリーダーシップスキルの向上を目指す手法を紹介。
012,コンピュータネットワークの基礎,西村さゆり,技術書典,3500,2022,コンピュータネットワークの基本概念から、実践的なネットワーク設計までを解説。
013,ビジネスプロセス最適化,小林太郎,ビジネスブックス,4600,2019,ビジネスプロセスの最適化手法とツールを駆使して企業の効率を向上させる手法を詳細に解説。
014,データウェアハウス設計入門,川上美香,データ出版社,5000,2021,データウェアハウスの基本概念から、実践的な設計手法までを解説。
015,マネジャーのためのコーチング,森田健太,ビジネスパートナーズ,3900,2018,リーダーシップとコーチングの基本原則と実践的なスキル向上のための手法を解説。
016,組み込みシステム開発ガイド,伊藤美和,技術出版株式会社,5500,2019,組み込みシステムの開発手法とデバッグテクニックに焦点を当てた実践的なガイド。
17,ディープラーニング実践入門,岡田淳,技術出版株式会社,4800,2022,ディープラーニングの基本理論から実践的な応用までを解説。
18,ビジネスプロセス改善ガイド,松本美香,ビジネスブックス,3500,2021,効果的なビジネスプロセス改善手法とツールを紹介。
19,Pythonデータサイエンス,田中健太,技術書典,4200,2020,Pythonを使用したデータサイエンスの基本と実践手法を学ぶ。
20,金融工学入門,林美香,データ出版社,5500,2019,金融商品の設計と価格付けに関する基本原則と数学的手法の解説。
21,チームリーダーシップの手引き,佐藤健太,ビジネスパートナーズ,4000,2020,効果的なチームリーダーシップの構築と維持に関するガイド。
22,React.js開発入門,伊藤さやか,技術書典,3800,2022,React.jsを使用したモダンなWebアプリケーションの開発手法を学ぶ。
23,戦略的マーケティング,山口直樹,ビジネスブックス,4500,2018,市場分析と戦略の構築に焦点を当てたマーケティング戦略のガイド。
24,機械学習プロジェクト管理,田村徹,技術出版株式会社,4900,2019,機械学習プロジェクトの計画と実行に関する手法を解説。
25,投資戦略の戦術,高橋優一,金融情報出版,3600,2021,リスク管理と効果的な投資戦略の構築に焦点を当てた投資戦略の戦術を解説。
26,サイバーセキュリティ対策,中村あや,技術書典,4300,2021,サイバーセキュリティの基本概念と対策手法を解説。
27,効果的なプロジェクトマネジメント,大塚和夫,ビジネスパートナーズ,4100,2020,プロジェクトの計画と実行に関する基本原則と実践的な手法を学ぶ。
28,データウェアハウス設計と実践,川上美香,データ出版社,5200,2021,データウェアハウスの設計と実践的なデータ統合手法を解説。
29,マーケットリサーチ入門,森田健太,ビジネスブックス,3700,2018,効果的なマーケットリサーチの実施とデータ分析手法を学ぶ。
30,クラウドセキュリティベストプラクティス,西村さゆり,技術出版株式会社,4700,2019,クラウド環境でのセキュリティのベストプラクティスを解説。
31,データサイエンティストのための統計学,小林太郎,データ書典,4400,2020,データサイエンスにおける統計学の基本と高度な手法までを解説。
32,リーダーシップ開発プログラム,中村美和,ビジネスパートナーズ,3900,2021,リーダーシップスキルの開発とチームビルディングの手法を学ぶ。
33,モバイルアプリケーション開発,高橋健一,技術書典,5000,2019,モバイルアプリケーションの開発手法とデザイン原則に焦点を当てた実践的なガイド。
34,投資ポートフォリオ戦略,田村美香,金融情報出版,4100,2020,リスク管理と効果的なポートフォリオの構築に焦点を当てた投資戦略の構築手法を解説。
35,サイバーセキュリティ実践入門,伊藤健太,技術書典,4600,2021,サイバーセキュリティの基本から、実践的な対策手法までを詳細に解説。
36,量子コンピューティング入門,岡本太郎,技術書典,5800,2022,量子コンピューティングの基礎理論から実践的な応用までを解説。
37,組織開発と変革マネジメント,村田花子,ビジネスブックス,4200,2021,組織の成熟度と変革マネジメントの手法に焦点を当てた実践的なガイド。
38,Javaマイクロサービスアーキテクチャ,佐々木健太,技術出版株式会社,4800,2020,Javaを使用したマイクロサービスアーキテクチャの基本と設計手法を学ぶ。
39,新時代のプロダクトマネジメント,林さやか,ビジネスパートナーズ,5500,2019,デジタルプロダクトのマネジメントにおける新たな手法とアプローチを解説。
40,ネットワークセキュリティ実践ガイド,高橋徹,技術書典,4600,2020,ネットワークセキュリティの実践的な対策手法とセキュアなネットワーク構築手法を学ぶ。
41,スタートアップのための財務戦略,伊東美和,ビジネスブックス,4000,2021,スタートアップ企業の財務戦略と資金調達の手法に焦点を当てたガイド。
42,自然言語処理と機械翻訳,山岸直樹,技術出版株式会社,5300,2018,自然言語処理と機械翻訳の基本原則と応用手法を学ぶ。
43,企業価値の最大化,佐野美香,データ出版社,4900,2019,企業価値の向上と最大化に向けた戦略と実践手法を解説。
44,データサイエンスとビジネス分析,鈴木健一,金融情報出版,3600,2020,データサイエンスとビジネスアナリティクスの基本概念と手法を学ぶ。
45,クラウドセキュリティポリシー,中川美和,技術書典,4400,2021,クラウド環境でのセキュリティポリシーの設計と実践手法に焦点を当てた実践ガイド。
46,リーダーシップ心理学,大西美香,ビジネスパートナーズ,4100,2018,リーダーシップと心理学の関連性と、組織内での実践手法を解説。
47,フロントエンド開発ベストプラクティス,田島健太,技術出版株式会社,4900,2019,モダンなフロントエンド開発のベストプラクティスと設計原則に焦点を当てた実践的なガイド。
48,投資リスク管理戦略,堀田直樹,金融情報出版,4200,2020,効果的な投資リスク管理戦略とポートフォリオの構築手法を解説。
49,セキュアなソフトウェア開発,大久保美和,技術書典,5500,2021,セキュアなソフトウェアの設計と実装に関する基本原則と手法を学ぶ。
50,ビジネスインテリジェンス基礎,小川美香,データ出版社,4700,2020,ビジネスインテリジェンスの基本的な原則とデータ分析手法を解説。
51,デジタルマーケティング入門,石田健太,ビジネスブックス,4000,2022,デジタル時代におけるマーケティング戦略と実践的なデジタル広告手法に焦点を当てた入門書。
52,モバイルセキュリティベストプラクティス,中山さやか,技術書典,4400,2022,モバイルアプリケーションのセキュリティ設計と対策手法を解説。
53,ビジネスプランの作成法,渡辺徹,ビジネスパートナーズ,3800,2019,効果的なビジネスプランの作成と実行に向けた手法を学ぶ。
54,クラウドネイティブアーキテクチャ,吉田太郎,技術出版株式会社,5200,2022,クラウドネイティブなアプリケーションの設計と構築に関する包括的なガイド。
55,組織学習の手法,石井花子,ビジネスブックス,4300,2021,組織が持続的に学習し成長するための手法と実践例を解説。
56,JavaScriptデザインパターン,小林次郎,技術書典,4600,2020,JavaScriptにおけるデザインパターンと実践的なプログラミング手法を学ぶ。
57,デジタルトランスフォーメーション戦略,鈴木美和,ビジネスパートナーズ,5400,2019,組織全体でのデジタルトランスフォーメーション戦略の策定と実行に関するガイド。
58,ネットワークプログラミング実践,田村健太,技術書典,4800,2020,ネットワークプログラミングの基本概念から、実践的なアプリケーション開発までを解説。
59,マーケットセグメンテーション,山本直樹,ビジネスブックス,4000,2021,効果的なマーケットセグメンテーションの手法と市場分析に焦点を当てた入門書。
60,サーバレスアーキテクチャ設計,田中美香,技術出版株式会社,5000,2018,サーバレスなアプリケーションの設計とデプロイに関する手法を解説。
61,人材育成プログラム,高橋直樹,ビジネスパートナーズ,3700,2019,組織内での効果的な人材育成プログラムの構築と実践手法を学ぶ。
62,Python機械学習実践ガイド,西村美和,技術書典,5300,2021,Pythonを使用した機械学習の基本理論から実践的な応用までを解説。
63,金融市場データ分析,堀田太郎,データ出版社,4600,2022,金融市場データの分析手法と取引戦略に焦点を当てた実践ガイド。
64,サイバーセキュリティリーダーシップ,矢田美香,技術出版株式会社,4800,2019,組織全体でのサイバーセキュリティリーダーシップの原則と実践手法を解説。
65,効果的なマーケティングコミュニケーション,小野次郎,ビジネスブックス,4200,2020,ターゲットに向けた効果的なマーケティングコミュニケーション戦略を解説。
66,データエンジニアリング入門,松井あや,技術書典,4900,2021,データエンジニアリングの基本概念と実践的なデータ処理手法を学ぶ。
67,戦略的イノベーションマネジメント,大木徹,ビジネスパートナーズ,5500,2020,企業における戦略的なイノベーションマネジメントの手法と実践例を解説。
68,ソフトウェアアーキテクチャ設計,藤田健一,技術出版株式会社,5100,2019,ソフトウェアアーキテクチャの基本原則と設計手法に焦点を当てた実践的なガイド。
69,投資ポートフォリオ最適化,石井さやか,金融情報出版,4300,2020,リスクとリターンのバランスを考慮した投資ポートフォリオ最適化の手法を解説。
70,マイクロサービスアーキテクチャ実践,森下太郎,技術出版株式会社,5200,2022,マイクロサービスアーキテクチャの実践的な設計と運用に関するガイド。
71,戦略的リーダーシップ,大橋美和,ビジネスブックス,4500,2021,組織戦略とリーダーシップの統合に焦点を当てた戦略的リーダーシップのガイド。
72,データウェアハウス最適化,田口健太,技術書典,4700,2020,データウェアハウスの最適化手法とパフォーマンスチューニングに関する実践的な知識。
73,プログラマのためのアルゴリズム,小山直樹,技術出版株式会社,3800,2019,プログラマ向けの基本的なアルゴリズムとデータ構造に焦点を当てた解説書。
74,デジタル戦略の構築,鈴木花子,ビジネスパートナーズ,5100,2019,デジタル時代におけるビジネス戦略の構築手法と実践例を解説。
75,ネットワークセキュリティベストプラクティス,高田徹,技術書典,4400,2022,ネットワークセキュリティのベストプラクティスと脅威対策に関するガイド。
76,人材評価と報酬戦略,小林美和,ビジネスブックス,4300,2021,組織内での効果的な人材評価と報酬戦略の構築手法を学ぶ。
77,リアクティブプログラミング入門,西野健太,技術出版株式会社,4800,2020,リアクティブなアプリケーション開発の基本理論と実践手法を解説。
78,データサイエンスと医療,大谷花子,データ出版社,5400,2021,医療データの解析とデータサイエンスの応用に焦点を当てた入門書。
79,クラウドアーキテクト設計,佐々木徹,技術書典,5000,2022,クラウドシステムのアーキテクト設計とベストプラクティスに関するガイド。
80,ビジネスエシックス,中川美和,ビジネスパートナーズ,3700,2020,ビジネス環境における倫理とエシックスの基本原則と実践的なガイド。
81,機械学習モデル解釈,高木健太,技術書典,4600,2019,機械学習モデルの解釈手法とモデルの透明性向上に焦点を当てた実践書。
82,セキュアなWeb開発,石崎さやか,技術出版株式会社,4300,2021,セキュアなWebアプリケーションの開発手法とセキュリティ対策に関するガイド。
83,デジタルマーケティング戦略,小野徹,ビジネスブックス,4900,2020,デジタルマーケティング戦略の立案と実行に関する実践的な手法を解説。
84,サーバーサイド開発入門,伊東美香,技術書典,4100,2022,サーバーサイド開発の基本概念から、実践的なアプリケーション開発までを解説。
85,データアーキテクチャ設計,松永直樹,データ出版社,5200,2019,データアーキテクチャの設計と実践的なデータ管理手法を学ぶ。
86,クラウドセキュリティアーキテクチャ,田辺美香,技術出版株式会社,5300,2022,クラウド環境におけるセキュリティアーキテクチャの設計と実践手法を解説。
87,リーダーシップ開発プログラム,小野さやか,ビジネスブックス,4100,2021,リーダーシップスキルの開発とチームビルディングの手法を学ぶ。
88,自然言語処理とテキストマイニング,村上健太,技術書典,4700,2020,自然言語処理とテキストマイニングの基本原則と実践手法を学ぶ。
89,デジタル戦略の実践,渡辺花子,ビジネスパートナーズ,5500,2019,デジタル時代におけるビジネス戦略の実践的な構築と実行に関するガイド。
90,フロントエンド開発ベストプラクティス,石井美和,技術出版株式会社,4800,2020,モダンなフロントエンド開発のベストプラクティスと設計原則に焦点を当てた実践的なガイド。
91,イノベーションマネジメント,佐々木太郎,ビジネスブックス,4300,2021,組織内での効果的なイノベーションマネジメントの手法と実践例を解説。
92,Pythonデータサイエンティスト養成,矢田さやか,技術出版株式会社,5100,2018,Pythonを使用したデータサイエンスの基本から実践までの養成ガイド。
93,金融市場リスク管理,山口健太,データ出版社,4900,2019,金融市場におけるリスク管理の基本原則と実践手法を解説。
94,マーケティングデータ分析,大塚美香,金融情報出版,4200,2020,効果的なマーケティングデータ分析の手法と意思決定への応用に焦点を当てた実践ガイド。
95,サイバーセキュリティリスクマネジメント,小川さやか,技術書典,4600,2022,サイバーセキュリティリスクの評価とマネジメントの手法に焦点を当てたガイド。
96,戦略的マーケティングプラン,松本美和,ビジネスパートナーズ,4700,2021,組織のビジョンと目標に基づく戦略的なマーケティングプランの構築手法を解説。
97,データサイエンスとビッグデータ,石田美香,技術書典,5300,2020,データサイエンスとビッグデータの基本原則から実践までの手法を学ぶ。
98,アジャイル開発プラクティス,吉田花子,技術出版株式会社,4000,2021,アジャイルなソフトウェア開発のプラクティスとスクラムの導入手法を解説。
99,サプライチェーンマネジメント,中村健太,データ出版社,5200,2022,効果的なサプライチェーンマネジメントの原則と実践手法を解説。
100,量子コンピュータ応用,田島美和,技術書典,4900,2020,量子コンピュータの応用領域と実践的な利用事例に焦点を当てたガイド。
101,エンタープライズアーキテクチャ,小林花子,ビジネスパートナーズ,4400,2019,エンタープライズ全体のアーキテクチャ設計と最適化手法に関する実践ガイド。
102,プロジェクトポートフォリオマネジメント,松井直樹,技術出版株式会社,5000,2020,プロジェクトポートフォリオの戦略的なマネジメント手法と実践例を解説。
103,マネジメント情報システム,大木美和,データブックス,4600,2021,組織内でのマネジメント情報システムの設計と運用に焦点を当てたガイド。

Writing books.csv


In [14]:
!head books.csv

ID,title,author,publisher,price,year,description
001,ネットワークセキュリティ入門,山田太郎,技術書典,4500,2021,ネットワークセキュリティの基本原則と実践的な手法に焦点を当てた入門書。
002,ビジネスデータ分析の基礎,田中花子,ビジネスブックス,3800,2020,ビジネスにおけるデータ分析の基本的な手法とツールの使い方を解説。
003,クラウドアーキテクチャデザイン,鈴木次郎,技術出版株式会社,5500,2022,クラウドネイティブなアプリケーションのアーキテクチャデザインとベストプラクティスを詳細に解説。
004,金融取引のアルゴリズム,林美香,データ出版社,5200,2019,金融市場におけるアルゴリズムトレーディングの基本原則と実践手法に焦点を当てた実践的なガイド。
005,人事マネジメントの実践,佐藤健太,ビジネスパートナーズ,4800,2020,組織内の人材マネジメントに関する基本的な原則と実践的な手法を紹介。
006,自然言語処理入門,伊藤さやか,技術書典,4200,2021,自然言語処理の基本概念から、実践的なテキストデータ処理手法までを解説。
007,マーケティング戦略の立案,山口直樹,ビジネスブックス,3600,2018,市場分析、セグメンテーション、ポジショニングなど、マーケティング戦略の構築手法を学ぶ。
008,ディープラーニングアプリケーション,田村徹,技術出版株式会社,6500,2019,ディープラーニングを使用した実践的なアプリケーション開発手法に焦点を当てた実践書。
009,投資ポートフォリオ管理,高橋優一,金融情報出版,3700,2020,効果的な投資ポートフォリオの構築と管理に関する手法を解説。


In [15]:
import pandas as pd

df = pd.read_csv("books.csv")
print(df)
print(type(df))

      ID                title author   publisher  price  year  \
0      1       ネットワークセキュリティ入門   山田太郎        技術書典   4500  2021   
1      2         ビジネスデータ分析の基礎   田中花子    ビジネスブックス   3800  2020   
2      3      クラウドアーキテクチャデザイン   鈴木次郎    技術出版株式会社   5500  2022   
3      4          金融取引のアルゴリズム    林美香      データ出版社   5200  2019   
4      5          人事マネジメントの実践   佐藤健太  ビジネスパートナーズ   4800  2020   
..   ...                  ...    ...         ...    ...   ...   
98    99       サプライチェーンマネジメント   中村健太      データ出版社   5200  2022   
99   100           量子コンピュータ応用   田島美和        技術書典   4900  2020   
100  101      エンタープライズアーキテクチャ   小林花子  ビジネスパートナーズ   4400  2019   
101  102  プロジェクトポートフォリオマネジメント   松井直樹    技術出版株式会社   5000  2020   
102  103         マネジメント情報システム   大木美和     データブックス   4600  2021   

                                         description  
0                ネットワークセキュリティの基本原則と実践的な手法に焦点を当てた入門書。  
1                   ビジネスにおけるデータ分析の基本的な手法とツールの使い方を解説。  
2    クラウドネイティブなアプリケーションのアーキテクチャデザインとベストプラクティスを詳細に解説。  

## Enbeddings Preperation

### Target Selection


In [16]:
descriptions = df["description"]

In [17]:
type(descriptions)

pandas.core.series.Series

In [18]:
targets = descriptions.tolist()
type(targets)

list

### LangChain Preparation

In [19]:
import langchain
from langchain.agents.agent_toolkits import create_retriever_tool, create_conversational_retrieval_agent
from langchain.chat_models import ChatOpenAI
from langchain.embeddings import OpenAIEmbeddings
from langchain.callbacks import StreamlitCallbackHandler
from langchain.schema import BaseRetriever, Document, SystemMessage

In [21]:
from langchain.globals import set_debug, set_verbose

set_debug(True)
set_verbose(False)

### OpenAI

In [51]:
import os
import openai
import getpass

os.environ["OPENAI_API_KEY"] = getpass.getpass()
#openai.api_key = os.getenv("OPENAI_API_KEY")


··········


In [52]:
import openai
from openai import OpenAI

client = OpenAI()
client.models.list()

SyncPage[Model](data=[Model(id='gpt-4-0613', created=1686588896, object='model', owned_by='openai'), Model(id='curie-search-query', created=1651172509, object='model', owned_by='openai-dev'), Model(id='gpt-4', created=1687882411, object='model', owned_by='openai'), Model(id='babbage-search-query', created=1651172509, object='model', owned_by='openai-dev'), Model(id='gpt-4-1106-preview', created=1698957206, object='model', owned_by='system'), Model(id='babbage-search-document', created=1651172510, object='model', owned_by='openai-dev'), Model(id='dall-e-2', created=1698798177, object='model', owned_by='system'), Model(id='gpt-3.5-turbo-0301', created=1677649963, object='model', owned_by='openai'), Model(id='gpt-3.5-turbo-16k-0613', created=1685474247, object='model', owned_by='openai'), Model(id='text-embedding-ada-002', created=1671217299, object='model', owned_by='openai-internal'), Model(id='davinci-search-query', created=1651172505, object='model', owned_by='openai-dev'), Model(id='

# Part A: カスタム Retriever利用

## DDL

In [38]:
session.execute("""
DROP TABLE IF EXISTS book_openai;
        """
)

In [39]:
session.execute("""DROP INDEX IF EXISTS idx_price_book_openai;""")
session.execute("""DROP INDEX IF EXISTS idx_year_book_openai;""")
session.execute("""DROP INDEX IF EXISTS idx_sem_vec_book_openai;""")
session.execute("""DROP TABLE IF EXISTS book_openai;""")

In [24]:
session.execute("""
CREATE TABLE  book_openai (
id text,
title text,
author text,
publisher text,
price int,
year int,
description text,
sem_vec vector<float, 1536>,
PRIMARY KEY((author, publisher), title, id)
);
        """
)

In [40]:
session.execute("""
CREATE TABLE  book_hface (
id text,
title text,
author text,
publisher text,
price int,
year int,
description text,
sem_vec vector<float, 768>,
PRIMARY KEY((author, publisher), title, id)
);
        """
)

In [ ]:
session.execute("CREATE CUSTOM INDEX idx_price_book_openai ON book_openai(price) USING 'org.apache.cassandra.index.sai.StorageAttachedIndex'")
session.execute("CREATE CUSTOM INDEX idx_year_book_openai ON book_openai(year) USING 'org.apache.cassandra.index.sai.StorageAttachedIndex'")
session.execute("CREATE CUSTOM INDEX idx_sem_vec_book_openai ON book_openai(sem_vec) USING 'org.apache.cassandra.index.sai.StorageAttachedIndex'")

In [41]:
session.execute("CREATE CUSTOM INDEX idx_price_book_hface ON book_hface(price) USING 'org.apache.cassandra.index.sai.StorageAttachedIndex'")
session.execute("CREATE CUSTOM INDEX idx_year_book_hface ON book_hface(year) USING 'org.apache.cassandra.index.sai.StorageAttachedIndex'")
session.execute("CREATE CUSTOM INDEX idx_sem_vec_book_hface ON book_hface(sem_vec) USING 'org.apache.cassandra.index.sai.StorageAttachedIndex'")

## Data Load

In [ ]:
embeddings = OpenAIEmbeddings()

In [27]:
!pip install sentence_transformers==2.2.2

In [28]:
from langchain.embeddings import HuggingFaceEmbeddings

# テキストの準備
text = "これは、テストドキュメントです。"

# HuggingFaceによる埋め込み生成
embeddings = HuggingFaceEmbeddings()
query_result = embeddings.embed_query(text)
doc_result = embeddings.embed_documents([text])

# 確認
print(len(query_result))
print(query_result)
print(len(doc_result[0]))
print(doc_result)

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:88: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


.gitattributes:   0%|          | 0.00/1.18k [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.6k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/571 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

data_config.json:   0%|          | 0.00/39.3k [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/438M [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/239 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/363 [00:00<?, ?B/s]

train_script.py:   0%|          | 0.00/13.1k [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

768
[0.03662785887718201, -0.02417769283056259, -0.026391828432679176, 0.020790087059140205, 0.03521691635251045, 0.013033317402005196, -0.0628475472331047, 0.07296635955572128, -0.008844171650707722, 0.004833828657865524, -0.010987279936671257, 0.00915275327861309, 0.03147989511489868, 0.009160381741821766, -0.04536428675055504, -0.08133849501609802, -0.012768767774105072, 0.019062096253037453, 0.02703108638525009, -0.013312385417521, 0.010177231393754482, -0.003451943164691329, -0.004237261135131121, -0.025030968710780144, 0.019513120874762535, -0.05111786723136902, 0.024503741413354874, -0.05386921763420105, 0.05552544817328453, 0.004435090813785791, 0.07280729711055756, -0.022053230553865433, -0.016681339591741562, -0.04687359556555748, 1.8266343886352843e-06, -0.009285438805818558, 0.015761494636535645, -0.007998165674507618, -0.026347890496253967, -0.009326200000941753, -0.04238429293036461, 0.002681078389286995, -0.00022174452897161245, 0.015793167054653168, -0.02591713890433311

In [29]:
openai_embedding_params = []
for target in targets:
  openai_embedding_params.append(embeddings.embed_query(target))

In [30]:
print(len(openai_embedding_params[0]))

768


In [31]:
haggingface_embedding_params = embeddings.embed_documents(targets)

In [34]:
print(len(haggingface_embedding_params[0]))

768


In [36]:
print(len(targets))
print(len(haggingface_embedding_params))

103
103


In [37]:
#df_openai = df.assign(sem_vec_openai=openai_embedding_params)
df_openai = df.assign(sem_vec=haggingface_embedding_params)
df_openai

ID                title author   publisher  price  year  \
0      1       ネットワークセキュリティ入門   山田太郎        技術書典   4500  2021   
1      2         ビジネスデータ分析の基礎   田中花子    ビジネスブックス   3800  2020   
2      3      クラウドアーキテクチャデザイン   鈴木次郎    技術出版株式会社   5500  2022   
3      4          金融取引のアルゴリズム    林美香      データ出版社   5200  2019   
4      5          人事マネジメントの実践   佐藤健太  ビジネスパートナーズ   4800  2020   
..   ...                  ...    ...         ...    ...   ...   
98    99       サプライチェーンマネジメント   中村健太      データ出版社   5200  2022   
99   100           量子コンピュータ応用   田島美和        技術書典   4900  2020   
100  101      エンタープライズアーキテクチャ   小林花子  ビジネスパートナーズ   4400  2019   
101  102  プロジェクトポートフォリオマネジメント   松井直樹    技術出版株式会社   5000  2020   
102  103         マネジメント情報システム   大木美和     データブックス   4600  2021   

                                         description  \
0                ネットワークセキュリティの基本原則と実践的な手法に焦点を当てた入門書。   
1                   ビジネスにおけるデータ分析の基本的な手法とツールの使い方を解説。   
2    クラウドネイティブなアプリケーションのアーキテクチャデザインとベストプラクティスを詳細に解説。   
3     金融市場におけるアルゴリズムトレーディングの基本原則と実践手法に焦点を当てた実践的なガイド。   
4                  組織内の人材マネジメントに関する基本的な原則と実践的な手法を紹介。   
..                                               ...   
98                    効果的なサプライチェーンマネジメントの原則と実践手法を解説。   
99                 量子コンピュータの応用領域と実践的な利用事例に焦点を当てたガイド。   
100             エンタープライズ全体のアーキテクチャ設計と最適化手法に関する実践ガイド。   
101               プロジェクトポートフォリオの戦略的なマネジメント手法と実践例を解説。   
102               組織内でのマネジメント情報システムの設計と運用に焦点を当てたガイド。   

                                               sem_vec  
0    [0.024106020107865334, 0.002243891591206193, -...  
1    [0.019792361184954643, -0.0321882963180542, -0...  
2    [0.02505706250667572, 0.007935326546430588, 0....  
3    [0.004268913064152002, 0.00947908777743578, -0...  
4    [0.0342194028198719, -0.004347567446529865, -0...  
..                                                 ...  
98   [0.031015705317258835, -0.028873931616544724, ...  
99   [0.012202728539705276, -0.01924460008740425, -...  
100  [0.03868180513381958, -0.01839403808116913, -0...  
101  [0.05236094072461128, -0.05124589428305626, -0...  
102  [0.02033027447760105, -0.03741474449634552, -0...  

[103 rows x 8 columns]

In [42]:
#query = "INSERT INTO book_openai(id,title,author,publisher,price,year,description,sem_vec) VALUES (?,?,?,?,?,?,?,?)"
query = "INSERT INTO book_hface(id,title,author,publisher,price,year,description,sem_vec) VALUES (?,?,?,?,?,?,?,?)"#
prepared = session.prepare(query)

In [43]:
params_list = []
for index, row in df_openai.iterrows():
  #params = [str(row['ID']),row['title'],row['author'],row['publisher'],row['price'],row['year'],row['description'],row['sem_vec_openai']]
  params = [str(row['ID']),row['title'],row['author'],row['publisher'],row['price'],row['year'],row['description'],row['sem_vec']]
  params_list.append(params)

In [44]:

from cassandra.concurrent import execute_concurrent_with_args
execute_concurrent_with_args(session, prepared, params_list)

[ExecutionResult(success=True, result_or_exc=<cassandra.cluster.ResultSet object at 0x782727cd33d0>),
 ExecutionResult(success=True, result_or_exc=<cassandra.cluster.ResultSet object at 0x782727cd3b80>),
 ExecutionResult(success=True, result_or_exc=<cassandra.cluster.ResultSet object at 0x782727cd2ec0>),
 ExecutionResult(success=True, result_or_exc=<cassandra.cluster.ResultSet object at 0x782727cd3be0>),
 ExecutionResult(success=True, result_or_exc=<cassandra.cluster.ResultSet object at 0x782727cd3700>),
 ExecutionResult(success=True, result_or_exc=<cassandra.cluster.ResultSet object at 0x782727cd3160>),
 ExecutionResult(success=True, result_or_exc=<cassandra.cluster.ResultSet object at 0x782727cd3670>),
 ExecutionResult(success=True, result_or_exc=<cassandra.cluster.ResultSet object at 0x782727cd3730>),
 ExecutionResult(success=True, result_or_exc=<cassandra.cluster.ResultSet object at 0x782727cd37c0>),
 ExecutionResult(success=True, result_or_exc=<cassandra.cluster.ResultSet object a

In [45]:
txt = 'データサイエンスについての書籍'

emb_query = embeddings.embed_query(txt)
len(emb_query)

768

In [ ]:
from cassandra.query import SimpleStatement
query = SimpleStatement("""
SELECT
similarity_cosine(sem_vec, %s) as similarity,
title, author, publisher, price, description
FROM book_openai
WHERE price > %s
ORDER BY sem_vec ANN OF %s LIMIT 5;
"""
)
price = 5000

results = session.execute(query, (emb_query, price, emb_query))
for row in results:
  print(row)

Row(similarity=0.95691978931427, title='データサイエンスと医療', author='大谷花子', publisher='データ出版社', price=5400, description='医療データの解析とデータサイエンスの応用に焦点を当てた入門書。')
Row(similarity=0.9507706165313721, title='データサイエンスとビッグデータ', author='石田美香', publisher='技術書典', price=5300, description='データサイエンスとビッグデータの基本原則から実践までの手法を学ぶ。')
Row(similarity=0.9408786296844482, title='Pythonデータサイエンティスト養成', author='矢田さやか', publisher='技術出版株式会社', price=5100, description='Pythonを使用したデータサイエンスの基本から実践までの養成ガイド。')
Row(similarity=0.9366496801376343, title='ディープラーニングアプリケーション', author='田村徹', publisher='技術出版株式会社', price=6500, description='ディープラーニングを使用した実践的なアプリケーション開発手法に焦点を当てた実践書。')
Row(similarity=0.9365187883377075, title='データアーキテクチャ設計', author='松永直樹', publisher='データ出版社', price=5200, description='データアーキテクチャの設計と実践的なデータ管理手法を学ぶ。')


In [46]:
from cassandra.query import SimpleStatement
query = SimpleStatement("""
SELECT
similarity_cosine(sem_vec, %s) as similarity,
title, author, publisher, price, description
FROM book_hface
WHERE price > %s
ORDER BY sem_vec ANN OF %s LIMIT 5;
"""
)
price = 5000

results = session.execute(query, (emb_query, price, emb_query))
for row in results:
  print(row)

Row(similarity=0.9192150235176086, title='データサイエンスと医療', author='大谷花子', publisher='データ出版社', price=5400, description='医療データの解析とデータサイエンスの応用に焦点を当てた入門書。')
Row(similarity=0.893189549446106, title='データサイエンスとビッグデータ', author='石田美香', publisher='技術書典', price=5300, description='データサイエンスとビッグデータの基本原則から実践までの手法を学ぶ。')
Row(similarity=0.859066903591156, title='ディープラーニングアプリケーション', author='田村徹', publisher='技術出版株式会社', price=6500, description='ディープラーニングを使用した実践的なアプリケーション開発手法に焦点を当てた実践書。')
Row(similarity=0.8494737148284912, title='組み込みシステム開発ガイド', author='伊藤美和', publisher='技術出版株式会社', price=5500, description='組み込みシステムの開発手法とデバッグテクニックに焦点を当てた実践的なガイド。')
Row(similarity=0.8494127988815308, title='データアーキテクチャ設計', author='松永直樹', publisher='データ出版社', price=5200, description='データアーキテクチャの設計と実践的なデータ管理手法を学ぶ。')


In [ ]:
txt = '量子コンピュータ'

emb_query = embeddings.embed_query(txt)
#emb_query.shape
len(emb_query)
from cassandra.query import SimpleStatement
query = SimpleStatement("""
SELECT
similarity_cosine(sem_vec, %s) as similarity,
title, author, publisher, price, description
FROM book_openai
ORDER BY sem_vec ANN OF %s LIMIT 5;
"""
)

results = session.execute(query, (emb_query, emb_query))
for row in results:
  print(row)

Row(similarity=0.9494655728340149, title='量子コンピューティング入門', author='岡本太郎', publisher='技術書典', price=5800, description='量子コンピューティングの基礎理論から実践的な応用までを解説。')
Row(similarity=0.9434464573860168, title='量子コンピュータ応用', author='田島美和', publisher='技術書典', price=4900, description='量子コンピュータの応用領域と実践的な利用事例に焦点を当てたガイド。')
Row(similarity=0.9098066687583923, title='Python機械学習実践ガイド', author='西村美和', publisher='技術書典', price=5300, description='Pythonを使用した機械学習の基本理論から実践的な応用までを解説。')
Row(similarity=0.9097812175750732, title='クラウドセキュリティベストプラクティス', author='西村さゆり', publisher='技術出版株式会社', price=4700, description='クラウド環境でのセキュリティのベストプラクティスを解説。')
Row(similarity=0.9092555046081543, title='データウェアハウス最適化', author='田口健太', publisher='技術書典', price=4700, description='データウェアハウスの最適化手法とパフォーマンスチューニングに関する実践的な知識。')


In [61]:
txt = '量子コンピュータ'

emb_query = embeddings.embed_query(txt)
#emb_query.shape
len(emb_query)
from cassandra.query import SimpleStatement
query = SimpleStatement("""
SELECT
similarity_cosine(sem_vec, %s) as similarity,
title, author, publisher, price, description
FROM book_hface
ORDER BY sem_vec ANN OF %s LIMIT 5;
"""
)

results = session.execute(query, (emb_query, emb_query))
for row in results:
  print(row)

Row(similarity=0.9030876755714417, title='量子コンピューティング入門', author='岡本太郎', publisher='技術書典', price=5800, description='量子コンピューティングの基礎理論から実践的な応用までを解説。')
Row(similarity=0.9025694131851196, title='量子コンピュータ応用', author='田島美和', publisher='技術書典', price=4900, description='量子コンピュータの応用領域と実践的な利用事例に焦点を当てたガイド。')
Row(similarity=0.8397799134254456, title='クラウドアーキテクチャデザイン', author='鈴木次郎', publisher='技術出版株式会社', price=5500, description='クラウドネイティブなアプリケーションのアーキテクチャデザインとベストプラクティスを詳細に解説。')
Row(similarity=0.8260414600372314, title='効果的なマーケティングコミュニケーション', author='小野次郎', publisher='ビジネスブックス', price=4200, description='ターゲットに向けた効果的なマーケティングコミュニケーション戦略を解説。')
Row(similarity=0.8131824135780334, title='マイクロサービスアーキテクチャ実践', author='森下太郎', publisher='技術出版株式会社', price=5200, description='マイクロサービスアーキテクチャの実践的な設計と運用に関するガイド。')


## LangChain

In [67]:
from cassandra.cluster import Session
from cassandra.query import PreparedStatement
class HFaceBookRetriever(BaseRetriever):
    session: Session
    embedding: HuggingFaceEmbeddings
    lang: str = "Japanese"
    search_statement_ja: PreparedStatement = None

    class Config:
        arbitrary_types_allowed = True

    def get_relevant_documents(self, query):
        docs = []
        embeddingvector = self.embedding.embed_query(query)
        if self.search_statement_ja is None:
            self.search_statement_ja = self.session.prepare("""
                SELECT
                    id,
                    similarity_cosine(sem_vec, ?) as similarity,
                    title,
                    author,
                    publisher,
                    price,
                    description
                FROM book_hface
                ORDER BY sem_vec ANN OF ?
                LIMIT ?
                """)
        query = self.search_statement_ja
        results = self.session.execute(query, [embeddingvector, embeddingvector, 5])
        top_products = results._current_rows
        for r in top_products:
            if r.similarity > 0.90:
                docs.append(Document(
                    id=r.id,
                    page_content=r.title,
                    metadata={"id": r.id,
                            "title": r.title,
                            "author": r.author,
                            "publisher": r.publisher,
                            "description": r.description,
                            "price": r.price
                            }
                ))

        return docs

In [ ]:
from cassandra.cluster import Session
from cassandra.query import PreparedStatement
class AstraBookRetriever(BaseRetriever):
    session: Session
    embedding: OpenAIEmbeddings
    lang: str = "Japanese"
    search_statement_ja: PreparedStatement = None

    class Config:
        arbitrary_types_allowed = True

    def get_relevant_documents(self, query):
        docs = []
        embeddingvector = self.embedding.embed_query(query)
        if self.search_statement_ja is None:
            self.search_statement_ja = self.session.prepare("""
                SELECT
                    id,
                    similarity_cosine(sem_vec, ?) as similarity,
                    title,
                    author,
                    publisher,
                    price,
                    description
                FROM book_openai
                ORDER BY sem_vec ANN OF ?
                LIMIT ?
                """)
        query = self.search_statement_ja
        results = self.session.execute(query, [embeddingvector, embeddingvector, 5])
        top_products = results._current_rows
        for r in top_products:
            if r.similarity > 0.91:
                docs.append(Document(
                    id=r.id,
                    page_content=r.title,
                    metadata={"id": r.id,
                            "title": r.title,
                            "author": r.author,
                            "publisher": r.publisher,
                            "description": r.description,
                            "price": r.price
                            }
                ))

        return docs

In [68]:
def create_chatbot(lang: str, Session: session):
    print(f"Creating chatbot for {lang}...")
    llm = ChatOpenAI(temperature=0, streaming=True)
    #embedding = OpenAIEmbeddings()
    embedding = HuggingFaceEmbeddings(cache_folder="./")

    retriever = HFaceBookRetriever(
        session=session, embedding=embedding, lang=lang)
    retriever_tool = create_retriever_tool(
        retriever, "books_retrevier", "Useful when searching for books from a book store. Prices are in YEN.")
    system_message = """
    You are a customer service of a book store and you are asked to pick books for a customer.
    You must try to find books related to given questions first.
    You must use the books_retreiver.
    You must not provide any information other than books that you get from books_retriever.
    You should behave as a bookstore clerk.
    """
    if lang == "Japanese":
        system_message = f"{system_message} All the responses should be in Japanese language."
    message = SystemMessage(content=system_message)
    agent_executor = create_conversational_retrieval_agent(
        llm=llm, tools=[retriever_tool], system_message=message, verbose=True)
    return agent_executor


In [69]:
lang = "Japanese"
chatbot = create_chatbot(lang, session)

Creating chatbot for Japanese...


In [70]:
prompt = "量子コンピュータについての本を教えて"
result = chatbot.invoke({
            "input": prompt})
print(result["output"])

[chain/start] [1:chain:AgentExecutor] Entering Chain run with input:
{
  "input": "量子コンピュータについての本を教えて",
  "chat_history": []
}
[llm/start] [1:chain:AgentExecutor > 2:llm:ChatOpenAI] Entering LLM run with input:
{
  "prompts": [
    "System: \n    You are a customer service of a book store and you are asked to pick books for a customer.\n    You must try to find books related to given questions first.\n    You must use the books_retreiver.\n    You must not provide any information other than books that you get from books_retriever.\n    You should behave as a bookstore clerk.\n     All the responses should be in Japanese language.\nHuman: 量子コンピュータについての本を教えて"
  ]
}
[llm/end] [1:chain:AgentExecutor > 2:llm:ChatOpenAI] [1.59s] Exiting LLM run with output:
{
  "generations": [
    [
      {
        "text": "",
        "generation_info": {
          "finish_reason": "function_call"
        },
        "type": "ChatGeneration",
        "message": {
          "lc": 1,
          "type": "constru

[tool/end] [1:chain:AgentExecutor > 3:tool:books_retrevier] s] Exiting Tool run with output:
"[Document(page_content='量子コンピューティング入門', metadata={'id': '36', 'title': '量子コンピューティング入門', 'author': '岡本太郎', 'publisher': '技術書典', 'description': '量子コンピューティングの基礎理論から実践的な応用までを解説。', 'price': 5800}), Document(page_content='量子コンピュータ応用', metadata={'id': '100', 'title': '量子コンピュータ応用', 'author': '田島美和', 'publisher': '技術書典', 'description': '量子コンピュータの応用領域と実践的な利用事例に焦点を当てたガイド。', 'price': 4900})]"
[llm/start] [1:chain:AgentExecutor > 5:llm:ChatOpenAI] Entering LLM run with input:
{
  "prompts": [
    "System: \n    You are a customer service of a book store and you are asked to pick books for a customer.\n    You must try to find books related to given questions first.\n    You must use the books_retreiver.\n    You must not provide any information other than books that you get from books_retriever.\n    You should behave as a bookstore clerk.\n     All the responses should be in Japanese language.\nHuman: 量子コ

# Part B: LangChain VectorStoreクラスからRetriever作成

In [71]:
from langchain.vectorstores import Cassandra

In [80]:
embeddings

HuggingFaceEmbeddings(client=SentenceTransformer(
  (0): Transformer({'max_seq_length': 384, 'do_lower_case': False}) with Transformer model: MPNetModel 
  (1): Pooling({'word_embedding_dimension': 768, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False})
  (2): Normalize()
), model_name='sentence-transformers/all-mpnet-base-v2', cache_folder=None, model_kwargs={}, encode_kwargs={}, multi_process=False)

In [84]:
session.execute("""
DROP TABLE IF EXISTS book_langchain;
        """
)

In [85]:
TABLE_NAME = "book_langchain"
vectorstore = Cassandra(embeddings, session, YOUR_KEYSPACE, TABLE_NAME)

以下のテーブルが作成される。

```
CREATE TABLE langchain.book_langchain (
    row_id text PRIMARY KEY,
    attributes_blob text,
    body_blob text,
    metadata_s map<text, text>,
    vector vector<float, 1536>
) WITH additional_write_policy = '99p'
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.UnifiedCompactionStrategy'}
    AND compression = {'chunk_length_in_kb': '16', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair = 'BLOCKING'
    AND speculative_retry = '99p';
CREATE CUSTOM INDEX eidx_metadata_s_book_langchain ON langchain.book_langchain (entries(metadata_s)) USING 'org.apache.cassandra.index.sai.StorageAttachedIndex';
CREATE CUSTOM INDEX idx_vector_book_langchain ON langchain.book_langchain (vector) USING 'org.apache.cassandra.index.sai.StorageAttachedIndex';

```

In [73]:
df

ID                title author   publisher  price  year  \
0      1       ネットワークセキュリティ入門   山田太郎        技術書典   4500  2021   
1      2         ビジネスデータ分析の基礎   田中花子    ビジネスブックス   3800  2020   
2      3      クラウドアーキテクチャデザイン   鈴木次郎    技術出版株式会社   5500  2022   
3      4          金融取引のアルゴリズム    林美香      データ出版社   5200  2019   
4      5          人事マネジメントの実践   佐藤健太  ビジネスパートナーズ   4800  2020   
..   ...                  ...    ...         ...    ...   ...   
98    99       サプライチェーンマネジメント   中村健太      データ出版社   5200  2022   
99   100           量子コンピュータ応用   田島美和        技術書典   4900  2020   
100  101      エンタープライズアーキテクチャ   小林花子  ビジネスパートナーズ   4400  2019   
101  102  プロジェクトポートフォリオマネジメント   松井直樹    技術出版株式会社   5000  2020   
102  103         マネジメント情報システム   大木美和     データブックス   4600  2021   

                                         description  
0                ネットワークセキュリティの基本原則と実践的な手法に焦点を当てた入門書。  
1                   ビジネスにおけるデータ分析の基本的な手法とツールの使い方を解説。  
2    クラウドネイティブなアプリケーションのアーキテクチャデザインとベストプラクティスを詳細に解説。  
3     金融市場におけるアルゴリズムトレーディングの基本原則と実践手法に焦点を当てた実践的なガイド。  
4                  組織内の人材マネジメントに関する基本的な原則と実践的な手法を紹介。  
..                                               ...  
98                    効果的なサプライチェーンマネジメントの原則と実践手法を解説。  
99                 量子コンピュータの応用領域と実践的な利用事例に焦点を当てたガイド。  
100             エンタープライズ全体のアーキテクチャ設計と最適化手法に関する実践ガイド。  
101               プロジェクトポートフォリオの戦略的なマネジメント手法と実践例を解説。  
102               組織内でのマネジメント情報システムの設計と運用に焦点を当てたガイド。  

[103 rows x 7 columns]

In [74]:
id_series = df["ID"]
ids = id_series.to_list()
type(ids)

list

In [75]:
texts = targets
type(texts)

list

In [76]:
ids

[1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103]

In [77]:
df_for_meta = df.drop(['ID','description'], axis=1)
df_for_meta

title author   publisher  price  year
0         ネットワークセキュリティ入門   山田太郎        技術書典   4500  2021
1           ビジネスデータ分析の基礎   田中花子    ビジネスブックス   3800  2020
2        クラウドアーキテクチャデザイン   鈴木次郎    技術出版株式会社   5500  2022
3            金融取引のアルゴリズム    林美香      データ出版社   5200  2019
4            人事マネジメントの実践   佐藤健太  ビジネスパートナーズ   4800  2020
..                   ...    ...         ...    ...   ...
98        サプライチェーンマネジメント   中村健太      データ出版社   5200  2022
99            量子コンピュータ応用   田島美和        技術書典   4900  2020
100      エンタープライズアーキテクチャ   小林花子  ビジネスパートナーズ   4400  2019
101  プロジェクトポートフォリオマネジメント   松井直樹    技術出版株式会社   5000  2020
102         マネジメント情報システム   大木美和     データブックス   4600  2021

[103 rows x 5 columns]

In [78]:
metadatas = df_for_meta.to_dict(orient='records')
metadatas[0]

{'title': 'ネットワークセキュリティ入門',
 'author': '山田太郎',
 'publisher': '技術書典',
 'price': 4500,
 'year': 2021}

In [86]:
await vectorstore.aadd_texts(texts, metadatas)

['c742bc88c887469985d01e2301a729e0',
 'e3a1b31dd62f4521bb712cc9873f4324',
 '86f2afe8e75241dd8e387bcdc40ac92e',
 'b56fce30e6f14526b88fdc18317d4782',
 '51f1e2d3720c4dca843ecc710223a024',
 'e08ba8dbc632469a86109c1c31efaf66',
 'b4fd34b4a99949f7a63c8e7d475cc267',
 'd95b4416d02f42579bb1f7658e6efa83',
 'bf67a54b68204cd2a8eda55fa5f6a46c',
 '29df5b27baf34b36993596c39561ae50',
 '927a86c38b174a7a97af756208a54b0d',
 '46030fedcef440929730f147ae540dc7',
 'c81abe3c5ec6458caf9f9de96773eb83',
 '50d11e193652451898a0c6c92d71ba54',
 '73e0755e14cb49f493d3bd2daa352291',
 'eb47df84fef74f2dadbc70ab5878bc5f',
 '21fa748fef39439a8b8262e36137fba6',
 '3173149614a84874aac460b537d9b49a',
 '19f10b8f0e5b4cc8ae5d8746ce015c09',
 '25a8316a8c6f46579974c9b65147f5dd',
 'dcfaf805ab7c4a1c95df3fb83bb6cde5',
 'cf8e10d856614ca091b31ae1bef5d476',
 '4cc79354debe4a84949d9571064fb38e',
 'e6d1c4293b95486181b06d48ec7bbe35',
 '9f61c09991b24e8a8dc0f68718fa724b',
 '59540c7eb1a84bef81a2a9a2960b3a6c',
 'c76dfeaa78b74f9da960e7ba64625834',
 

```
token@cqlsh:langchain> select row_id from langchain.book_langchain limit 1;

 row_id
----------------------------------
 b3097475f6a64bd68a1b543440cf9e48

(1 rows)
token@cqlsh:langchain> select metadata_s from langchain.book_langchain limit 1;

 metadata_s
-------------------------------------------------------------------------------------------------------------------------------
 {'author': '伊藤健太', 'price': '4600.0', 'publisher': '技術書典', 'title': 'サイバーセキュリティ実践入門', 'year': '2021.0'}

(1 rows)
```

In [87]:
session.execute("""
DROP TABLE IF EXISTS books_with_id;
        """
)

In [88]:
TABLE_NAME_WITH_ID = "books_with_id"
vectorstore_with_id = Cassandra(embeddings, session, YOUR_KEYSPACE, TABLE_NAME_WITH_ID)

```
token@cqlsh:langchain> desc books_with_id;

CREATE TABLE langchain.books_with_id (
    row_id text PRIMARY KEY,
    attributes_blob text,
    body_blob text,
    metadata_s map<text, text>,
    vector vector<float, 1536>
) WITH additional_write_policy = '99p'
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.UnifiedCompactionStrategy'}
    AND compression = {'chunk_length_in_kb': '16', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair = 'BLOCKING'
    AND speculative_retry = '99p';
CREATE CUSTOM INDEX eidx_metadata_s_books_with_id ON langchain.books_with_id (entries(metadata_s)) USING 'org.apache.cassandra.index.sai.StorageAttachedIndex';
CREATE CUSTOM INDEX idx_vector_books_with_id ON langchain.books_with_id (vector) USING 'org.apache.cassandra.index.sai.StorageAttachedIndex';
```

In [89]:
str_ids = [str(i) for i in ids]

In [90]:
vectorstore_with_id.add_texts(texts, metadatas, str_ids)

['1',
 '2',
 '3',
 '4',
 '5',
 '6',
 '7',
 '8',
 '9',
 '10',
 '11',
 '12',
 '13',
 '14',
 '15',
 '16',
 '17',
 '18',
 '19',
 '20',
 '21',
 '22',
 '23',
 '24',
 '25',
 '26',
 '27',
 '28',
 '29',
 '30',
 '31',
 '32',
 '33',
 '34',
 '35',
 '36',
 '37',
 '38',
 '39',
 '40',
 '41',
 '42',
 '43',
 '44',
 '45',
 '46',
 '47',
 '48',
 '49',
 '50',
 '51',
 '52',
 '53',
 '54',
 '55',
 '56',
 '57',
 '58',
 '59',
 '60',
 '61',
 '62',
 '63',
 '64',
 '65',
 '66',
 '67',
 '68',
 '69',
 '70',
 '71',
 '72',
 '73',
 '74',
 '75',
 '76',
 '77',
 '78',
 '79',
 '80',
 '81',
 '82',
 '83',
 '84',
 '85',
 '86',
 '87',
 '88',
 '89',
 '90',
 '91',
 '92',
 '93',
 '94',
 '95',
 '96',
 '97',
 '98',
 '99',
 '100',
 '101',
 '102',
 '103']

```
token@cqlsh:langchain> select count(row_id) from langchain.books_with_id;

 system.count(row_id)
----------------------
                  103

(1 rows)
```

```
token@cqlsh:langchain> select row_id, metadata_s from books_with_id limit 1;

 row_id | metadata_s
--------+---------------------------------------------------------------------------------------------------------------------
      6 | {'author': '伊藤さやか', 'price': '4200.0', 'publisher': '技術書典', 'title': '自然言語処理入門', 'year': '2021.0'}

(1 rows)
token@cqlsh:langchain>
```

In [91]:
retriever_from_vectorstore = vectorstore_with_id.as_retriever()

In [92]:
def create_chatbot_with_retriever(lang: str, session, retriever):
    print(f"Creating chatbot for {lang}...")
    llm = ChatOpenAI(temperature=0, streaming=True)
    embedding = OpenAIEmbeddings()

    retriever_tool = create_retriever_tool(
        retriever, "book_retrevier", "Useful when searching for books from a book store. Prices are in YEN.")
    system_message = """
    You are a customer service of a book store and you are asked to pick books for a customer.
    You must try to find books related to given questions first.
    You must use the books_retreiver.
    You must not provide any information other than books that you get from books_retriever.
    You should behave as a bookstore clerk.
    """
    if lang == "Japanese":
        system_message = f"{system_message} All the responses should be in Japanese language."
    message = SystemMessage(content=system_message)
    agent_executor = create_conversational_retrieval_agent(
        llm=llm, tools=[retriever_tool], system_message=message, verbose=True)
    return agent_executor

In [93]:
lang = "Japanese"
chatbot_cassio = create_chatbot_with_retriever(lang, session, retriever_from_vectorstore)

Creating chatbot for Japanese...


/usr/local/lib/python3.10/dist-packages/langchain_core/_api/deprecation.py:117: LangChainDeprecationWarning: The class `langchain_community.embeddings.openai.OpenAIEmbeddings` was deprecated in langchain-community 0.1.0 and will be removed in 0.2.0. Use langchain_openai.OpenAIEmbeddings instead.
  warn_deprecated(


In [94]:
prompt = "機械学習についての本を教えて"
vectorstore_with_id.similarity_search(prompt)

[Document(page_content='データサイエンスにおける統計学の基本と高度な手法までを解説。', metadata={'author': '小林太郎', 'price': '4400.0', 'publisher': 'データ書典', 'title': 'データサイエンティストのための統計学', 'year': '2020.0'}),
 Document(page_content='データサイエンスとビジネスアナリティクスの基本概念と手法を学ぶ。', metadata={'author': '鈴木健一', 'price': '3600.0', 'publisher': '金融情報出版', 'title': 'データサイエンスとビジネス分析', 'year': '2020.0'}),
 Document(page_content='セキュアなソフトウェアの設計と実装に関する基本原則と手法を学ぶ。', metadata={'author': '大久保美和', 'price': '5500.0', 'publisher': '技術書典', 'title': 'セキュアなソフトウェア開発', 'year': '2021.0'}),
 Document(page_content='データサイエンスとビッグデータの基本原則から実践までの手法を学ぶ。', metadata={'author': '石田美香', 'price': '5300.0', 'publisher': '技術書典', 'title': 'データサイエンスとビッグデータ', 'year': '2020.0'})]

In [ ]:
prompt = "機械学習についての本を教えて"
result = chatbot_cassio.invoke({
            "input": prompt})
print(result["output"])

[chain/start] [1:chain:AgentExecutor] Entering Chain run with input:
{
  "input": "機械学習についての本を教えて",
  "chat_history": []
}
[llm/start] [1:chain:AgentExecutor > 2:llm:ChatOpenAI] Entering LLM run with input:
{
  "prompts": [
    "System: \n    You are a customer service of a book store and you are asked to pick books for a customer.\n    You must try to find books related to given questions first.\n    You must use the books_retreiver.\n    You must not provide any information other than books that you get from books_retriever.\n    You should behave as a bookstore clerk.\n     All the responses should be in Japanese language.\nHuman: 機械学習についての本を教えて"
  ]
}
[llm/end] [1:chain:AgentExecutor > 2:llm:ChatOpenAI] [2.45s] Exiting LLM run with output:
{
  "generations": [
    [
      {
        "text": "",
        "generation_info": {
          "finish_reason": "function_call"
        },
        "type": "ChatGenerationChunk",
        "message": {
          "lc": 1,
          "type": "constructo

[tool/end] [1:chain:AgentExecutor > 3:tool:book_retrevier] [1.35s] Exiting Tool run with output:
"[Document(page_content='機械学習プロジェクトの計画と実行に関する手法を解説。', metadata={'author': '田村徹', 'price': '4900.0', 'publisher': '技術出版株式会社', 'title': '機械学習プロジェクト管理', 'year': '2019.0'}), Document(page_content='Pythonを使用した機械学習の基本理論から実践的な応用までを解説。', metadata={'author': '西村美和', 'price': '5300.0', 'publisher': '技術書典', 'title': 'Python機械学習実践ガイド', 'year': '2021.0'}), Document(page_content='機械学習モデルの解釈手法とモデルの透明性向上に焦点を当てた実践書。', metadata={'author': '高木健太', 'price': '4600.0', 'publisher': '技術書典', 'title': '機械学習モデル解釈', 'year': '2019.0'}), Document(page_content='自然言語処理と機械翻訳の基本原則と応用手法を学ぶ。', metadata={'author': '山岸直樹', 'price': '5300.0', 'publisher': '技術出版株式会社', 'title': '自然言語処理と機械翻訳', 'year': '2018.0'})]"
[llm/start] [1:chain:AgentExecutor > 5:llm:ChatOpenAI] Entering LLM run with input:
{
  "prompts": [
    "System: \n    You are a customer service of a book store and you are asked to pick books for a customer.\n    You 

## 付録. 様々な検索メソッド

In [95]:
vectorstore_with_id.similarity_search_with_relevance_scores(prompt)

[(Document(page_content='データサイエンスにおける統計学の基本と高度な手法までを解説。', metadata={'author': '小林太郎', 'price': '4400.0', 'publisher': 'データ書典', 'title': 'データサイエンティストのための統計学', 'year': '2020.0'}),
  0.8979912170573654),
 (Document(page_content='データサイエンスとビジネスアナリティクスの基本概念と手法を学ぶ。', metadata={'author': '鈴木健一', 'price': '3600.0', 'publisher': '金融情報出版', 'title': 'データサイエンスとビジネス分析', 'year': '2020.0'}),
  0.8778598154689805),
 (Document(page_content='セキュアなソフトウェアの設計と実装に関する基本原則と手法を学ぶ。', metadata={'author': '大久保美和', 'price': '5500.0', 'publisher': '技術書典', 'title': 'セキュアなソフトウェア開発', 'year': '2021.0'}),
  0.8731020488526386),
 (Document(page_content='データサイエンスとビッグデータの基本原則から実践までの手法を学ぶ。', metadata={'author': '石田美香', 'price': '5300.0', 'publisher': '技術書典', 'title': 'データサイエンスとビッグデータ', 'year': '2020.0'}),
  0.8663834602887917)]

In [96]:
vectorstore_with_id.similarity_search_with_relevance_scores(prompt,filter={'publisher':'技術書典'})

[(Document(page_content='セキュアなソフトウェアの設計と実装に関する基本原則と手法を学ぶ。', metadata={'author': '大久保美和', 'price': '5500.0', 'publisher': '技術書典', 'title': 'セキュアなソフトウェア開発', 'year': '2021.0'}),
  0.8731020488526386),
 (Document(page_content='データサイエンスとビッグデータの基本原則から実践までの手法を学ぶ。', metadata={'author': '石田美香', 'price': '5300.0', 'publisher': '技術書典', 'title': 'データサイエンスとビッグデータ', 'year': '2020.0'}),
  0.8663834602887917),
 (Document(page_content='データエンジニアリングの基本概念と実践的なデータ処理手法を学ぶ。', metadata={'author': '松井あや', 'price': '4900.0', 'publisher': '技術書典', 'title': 'データエンジニアリング入門', 'year': '2021.0'}),
  0.8509537144468682),
 (Document(page_content='自然言語処理とテキストマイニングの基本原則と実践手法を学ぶ。', metadata={'author': '村上健太', 'price': '4700.0', 'publisher': '技術書典', 'title': '自然言語処理とテキストマイニング', 'year': '2020.0'}),
  0.8500573280646435)]

In [97]:
vectorstore_with_id.similarity_search_with_relevance_scores(prompt,filter={'author':'西村美和'})

[(Document(page_content='Pythonを使用した機械学習の基本理論から実践的な応用までを解説。', metadata={'author': '西村美和', 'price': '5300.0', 'publisher': '技術書典', 'title': 'Python機械学習実践ガイド', 'year': '2021.0'}),
  0.7992703734478201)]

In [98]:
vectorstore_with_id.similarity_search_with_relevance_scores(prompt,filter={'publisher':'技術書典', 'author':'西村美和'})

[(Document(page_content='Pythonを使用した機械学習の基本理論から実践的な応用までを解説。', metadata={'author': '西村美和', 'price': '5300.0', 'publisher': '技術書典', 'title': 'Python機械学習実践ガイド', 'year': '2021.0'}),
  0.7992703734478201)]

In [99]:
vectorstore_with_id.similarity_search_with_relevance_scores(prompt,filter={'publisher':'技術書典', 'title':'自然言語'})

[]

In [100]:
vectorstore_with_id.max_marginal_relevance_search(prompt)

[Document(page_content='データサイエンスにおける統計学の基本と高度な手法までを解説。', metadata={'author': '小林太郎', 'price': '4400.0', 'publisher': 'データ書典', 'title': 'データサイエンティストのための統計学', 'year': '2020.0'}),
 Document(page_content='自然言語処理と機械翻訳の基本原則と応用手法を学ぶ。', metadata={'author': '山岸直樹', 'price': '5300.0', 'publisher': '技術出版株式会社', 'title': '自然言語処理と機械翻訳', 'year': '2018.0'}),
 Document(page_content='医療データの解析とデータサイエンスの応用に焦点を当てた入門書。', metadata={'author': '大谷花子', 'price': '5400.0', 'publisher': 'データ出版社', 'title': 'データサイエンスと医療', 'year': '2021.0'}),
 Document(page_content='ビジネス環境における倫理とエシックスの基本原則と実践的なガイド。', metadata={'author': '中川美和', 'price': '3700.0', 'publisher': 'ビジネスパートナーズ', 'title': 'ビジネスエシックス', 'year': '2020.0'})]

In [101]:
vectorstore_with_id.max_marginal_relevance_search(prompt, lambda_mult=0.1)

[Document(page_content='データサイエンスにおける統計学の基本と高度な手法までを解説。', metadata={'author': '小林太郎', 'price': '4400.0', 'publisher': 'データ書典', 'title': 'データサイエンティストのための統計学', 'year': '2020.0'}),
 Document(page_content='自然言語処理と機械翻訳の基本原則と応用手法を学ぶ。', metadata={'author': '山岸直樹', 'price': '5300.0', 'publisher': '技術出版株式会社', 'title': '自然言語処理と機械翻訳', 'year': '2018.0'}),
 Document(page_content='ビジネス環境における倫理とエシックスの基本原則と実践的なガイド。', metadata={'author': '中川美和', 'price': '3700.0', 'publisher': 'ビジネスパートナーズ', 'title': 'ビジネスエシックス', 'year': '2020.0'}),
 Document(page_content='リーダーシップと心理学の関連性と、組織内での実践手法を解説。', metadata={'author': '大西美香', 'price': '4100.0', 'publisher': 'ビジネスパートナーズ', 'title': 'リーダーシップ心理学', 'year': '2018.0'})]

In [102]:
vectorstore_with_id.max_marginal_relevance_search(prompt, lambda_mult=0.9)

[Document(page_content='データサイエンスにおける統計学の基本と高度な手法までを解説。', metadata={'author': '小林太郎', 'price': '4400.0', 'publisher': 'データ書典', 'title': 'データサイエンティストのための統計学', 'year': '2020.0'}),
 Document(page_content='データサイエンスとビジネスアナリティクスの基本概念と手法を学ぶ。', metadata={'author': '鈴木健一', 'price': '3600.0', 'publisher': '金融情報出版', 'title': 'データサイエンスとビジネス分析', 'year': '2020.0'}),
 Document(page_content='セキュアなソフトウェアの設計と実装に関する基本原則と手法を学ぶ。', metadata={'author': '大久保美和', 'price': '5500.0', 'publisher': '技術書典', 'title': 'セキュアなソフトウェア開発', 'year': '2021.0'}),
 Document(page_content='組織が持続的に学習し成長するための手法と実践例を解説。', metadata={'author': '石井花子', 'price': '4300.0', 'publisher': 'ビジネスブックス', 'title': '組織学習の手法', 'year': '2021.0'})]

In [103]:
# Retrieve more documents with higher diversity
# Useful if your dataset has many similar documents
retriever_a = vectorstore_with_id.as_retriever(
    search_type="mmr",
    search_kwargs={'k': 6, 'lambda_mult': 0.25}
)

# Fetch more documents for the MMR algorithm to consider
# But only return the top 5
retriever_b = vectorstore_with_id.as_retriever(
    search_type="mmr",
    search_kwargs={'k': 5, 'fetch_k': 50}
)

# Only retrieve documents that have a relevance score
# Above a certain threshold
retriever_c = vectorstore_with_id.as_retriever(
    search_type="similarity_score_threshold",
    search_kwargs={'score_threshold': 0.95}
)

# Only get the single most similar document from the dataset
retriever_d = vectorstore_with_id.as_retriever(search_kwargs={'k': 1})

# Use a filter to only retrieve documents from a specific paper
retriever_e = vectorstore_with_id.as_retriever(
    search_kwargs={'filter': {'paper_title':'GPT-4 Technical Report'}}
)